main/src/ca/uwaterloo/flix/api/lsp/provider/completion/KeywordCompleters.scala

```diff
@@ -0,0 +1,199 @@
+import ca.uwaterloo.flix.api.lsp.Index
+import ca.uwaterloo.flix.language.ast.TypedAst
+
+/**
+  * A collection of keyword completers, including
+  * 
+  * - `Decl`: completer for declaraiton keywords
+  * - `Enum`: completer for declaration keywords
+  * - `Expr`: completer for expressoin keywords
+  * - `Other`: completer for miscellaneous keywords 
+  * 
+  */
+object KeywordCompleters {
+
+    /**
+      * A completer for miscellaneous keywords.
+      */ 
+    object Other extends Completer {
+        def getCompletions(context: CompletionContext)(implicit flix: Flix, index: Index, root: TypedAst.Root): Iterable[Completion] =
+            List(
+            ("with", Priority.highest),
+            ("law", Priority.higher),
+            ("@Test", Priority.high),
+            ("where", Priority.low),
+            ("fix", Priority.low),
+            ("@Deprecated", Priority.lowest),
+            ("@Parallel", Priority.lowest),
+            ("@ParallelWhenPure", Priority.lowest),
+            ("@Lazy", Priority.lowest),
+            ("@LazyWhenPure", Priority.lowest),
+            ("Record", Priority.lowest),
+            ("redef", Priority.lowest),
+            ("Schema", Priority.lowest),
+            ) map { case (name, priority) => Completion.KeywordCompletion(name, priority(name)) }
+    }
+
+    /**
+      * A completer for trait declaration keywords. These are keywords that occur within a trait declaration.
+      */
+    object Trait extends Completer {
+      def getCompletions(context: CompletionContext)(implicit flix: Flix, index: Index, root: TypedAst.Root): Iterable[Completion] = {
+        if (context.previousWord == "pub") {
+          Completion.KeywordCompletion("def", Priority.lower("def")) :: Nil
+        } else {
+          Completion.KeywordCompletion("pub", Priority.lower("pub")) :: Nil
+        }
+      }
+    }
+
+
+    /**
+      * A completer for declaration keywords. These are keywords that denote a declaration.
+      */
+    object Decl extends Completer {
+      val DEF       = "def"
+      val PUB       = "pub"
+      val ENUM      = "enum"
+      val TYPE      = "type"
+      val INSTANCE  = "instance"
+      val MOD       = "mod"
+      val EFF       = "eff"
+      val STRUCT    = "struct"
+      val SEALED    = "sealed"
+      val TRAIT     = "trait"
+      val IMPORT    = "import"
+
+      val DEF_ENTRY       = (DEF      , Priority.highest)
+      val PUB_ENTRY       = (PUB      , Priority.higher)
+      val ENUM_ENTRY      = (ENUM     , Priority.high)
+      val TYPE_ENTRY      = (TYPE     , Priority.high)
+      val INSTANCE_ENTRY  = (INSTANCE , Priority.high)
+      val MOD_ENTRY       = (MOD      , Priority.low)
+      val EFF_ENTRY       = (EFF      , Priority.lower)
+      val STRUCT_ENTRY    = (STRUCT   , Priority.lower)
+      val SEALED_ENTRY    = (SEALED   , Priority.lowest)
+      val TRAIT_ENTRY     = (TRAIT    , Priority.lowest)
+      val IMPORT_ENTRY    = (IMPORT   , Priority.lowest)
```

Comment:
All of the above looks crazy :-)

Just write some functions that return the right thing. Don't introduce layers and layers of abstraction!